react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js

```diff
@@ -8,16 +8,31 @@ const initialState = {
 
 function reducer(state = initialState, action) {
   switch (action.type) {
-    case actions.GET_BOOKS: // TODO to implement the logic
-      return { ...state };
-    case actions.ADD_TO_CART: // TODO to implement the logic
-      return { ...state };
-    case actions.ADD_ITEM: // TODO to implement the logic
-      return { ...state };
-    case actions.REMOVE_ITEM: // TODO to implement the logic
-      return { ...state };
-    case actions.SEARCH_ITEM: // TODO to implement the logic
-      return { ...state };
+    case actions.GET_BOOKS:
+      return { ...state, books: action.payload, originalData: action.payload };
+    case actions.ADD_TO_CART:
+      return { ...state, bookSelected: [...state.bookSelected, action.payload] };
+    case actions.ADD_ITEM: {
+      const addItem = state.bookSelected.find(book => book.id === action.payload);
+      addItem.quantity += 1;
+      return { ...state, bookSelected: [...state.bookSelected] };
+    }
+    case actions.REMOVE_ITEM: {
+      const newBooksSelected = state.bookSelected.filter(book => book.id !== action.payload);
+      return { ...state, bookSelected: [...newBooksSelected] };
+    }
+    case actions.SUBTRACT_ITEM: {
+      const substractItem = state.bookSelected.find(book => book.id === action.payload);
+      substractItem.quantity -= 1;
+      return { ...state, bookSelected: [...state.bookSelected] };
+    }
+    case actions.SEARCH_ITEM: {
+      const books = [...state.originalData];
```

Comment:
this is the correct way to do this :) you can copy this logic into the other actions, here yoou are noot mnutating the state, instead you are working with a copy and return a new state :) nice!